### ISSUE TITLE
ST_ClusterDBSCAN implementation

### ISSUE THREAD

**Issue by baptistegirault** [duckdb/duckdb_spatial#412]
PostGIS ST_ClusterDBSCAN is a really powerfull function that is only available in PostGIS. This function return cluster number from a list of geometry based on DBSCAN.

https://postgis.net/docs/ST_ClusterDBSCAN.html

**Comment by VGSML** [comment]
Is there any plan to work on it? As well as on ST_ClusterIntersect and ST_ClusterKMeans?

**Comment by Maxxen** [comment]
Yes, I'm waiting for some changes in DuckDBs sorting and partitioning code to land, after which I will try to integrate these into spatial.

**Comment by stephensilber** [comment]
We're also hoping to use the ST_ClusterDBSCAN. @Maxxen any idea if this is something that is weeks away or are we looking at months/year out?

**Comment by Maxxen** [comment]
Hard to say for sure but Im currently reworking distance computations in general, after which a naive in-memory DBSCAN shouldn't be too hard to implement. The larger sorting and partitioning changes required to make it really performant are probably not going to be in place until next major release.

**Comment by bochkov-vi** [comment]
ST_ClusterDBSCAN is very lacking

**Comment by bsplosion** [comment]
Happen to have any updates here?  We're looking at a migration to DuckDB but we use this function pretty heavily with PostGIS today and there's not really any other equivalent of the algorithm.

**Comment by Maxxen** [comment]
Ive got ST_ClusterDBScan working from spatial's side, but it requires changes in DuckDB core in order to force the window function executor to always call the right callbacks. Im going to talk to the rest of the team to figure out how to proceed, and after that should be able to get it in for DuckDB v1.4 scheduled for September.

**Comment by fhk** [comment]
One thing to consider here is the ability to call partition also as alot of the time you're unable to call DBScan on larger areas.

**Comment by bsplosion** [comment]
> Ive got ST_ClusterDBScan working from spatial's side, but it requires changes in DuckDB core in order to force the window function executor to always call the right callbacks. Im going to talk to the rest of the team to figure out how to proceed, and after that should be able to get it in for DuckDB v1.4 scheduled for September.

That would super helpful!  There's no strong replacement algorithm available for us so having it in place on the next major release is exciting.

@fhk would there be a reason you can't call DBSCAN on larger areas?  It does have density/distance function contraints, of course, but we haven't really seen any issues with other implementations of DBSCAN, even when it comes to large datasets.  It should be able to run in O(n log n) time if implemented properly.

**Comment by fhk** [comment]
Ah ! Depends on your usecase and what you mean by big.

Alot of the time you want to cluster more than a million points.

My current method of choice for this is to use BigQuery or cuml.

Specifically running a whole US state or even a Country.

Typically there will be too much GEOMETRY to fit it all in memory. So people work around that by either pre partitioning the data or doing it in the query like so - https://stackoverflow.com/questions/68811457/clustering-using-dbscan-in-bigquery

**Comment by bsplosion** [comment]
> Ive got ST_ClusterDBScan working from spatial's side, but it requires changes in DuckDB core in order to force the window function executor to always call the right callbacks. Im going to talk to the rest of the team to figure out how to proceed, and after that should be able to get it in for DuckDB v1.4 scheduled for September.

Hi @Maxxen , any chance this actually landed in 1.4? I see that version was released today but this issue is still marked as open, so I’m guessing it didn’t make the cut.

**Comment by fhk** [comment]
just wanted to checkin :)

**Comment by bsplosion** [comment]
@Maxxen if you happen to have any updates on this I'd really appreciate it!  We're now a bit stuck on a migration because of a dependency on this algorithm without any strong substitutes available.  Is there a time where a merge is planned?

I'm not competent within the core so unfortunately I won't be of much help actually making that happen, but I'm 100% here for moral support!

_Edit_ I just checked the preview of the stable docs for the spatial extension and it's not currently in there as of 2026-02-03: https://duckdb.org/docs/preview/core_extensions/spatial/functions

**Comment by Maxxen** [comment]
Hi! 
This will not make it for v1.5. We're currently still focused on joins, statistics and compression, but once that is over I hope I can switch back to window functions for v1.6.

**Comment by bsplosion** [comment]
> Hi! This will not make it for v1.5. We're currently still focused on joins, statistics and compression, but once that is over I hope I can switch back to window functions for v1.6.

All right, it sounds like we really shouldn't rely on this definitely being available at any point necessarily.  We ended up having to hybridize a roughly-equivalent solution using a combination of H3 and a Python Union-Find algorithm run out of band, which somewhat works but is definitely not as robust as an actual `ST_ClusterDBSCAN` implementation would be.

**Comment by beckernick** [comment]
Since @fhk mentioned using cuML for GPU-accelerated clustering in a prior comment, figured I'd show a potential example of something like that in case folks are interested. (I work on accelerated data processing at NVIDIA).

If you have a GPU available, popping between DuckDB and cuML should hopefully be fairly smooth from a UX perspective.

(Slightly tweaked Gemini output):

```python
import duckdb
import pandas as pd
from cuml.cluster import HDBSCAN

# 1. Connect to DuckDB and create some dummy data
con = duckdb.connect()
con.execute("""
    CREATE TABLE points AS 
    SELECT
        random() AS x1,
        random() AS x2,
        random() AS y,
    FROM range(10000)
""")

# 2. Run your query and fetch to a DataFrame
query = "SELECT * FROM points"
table = con.execute(query).df()

# 3. Initialize and fit HDBSCAN (or DBSCAN)
hdbscan = HDBSCAN()
hdbscan.fit(table)
table["labels"] = hdbscan.labels_

# 4. Create a new table in DuckDB using the results
con.register('results', table)
con.execute("""
    CREATE TABLE points_clustered AS 
    SELECT * FROM results
""")

# Verify the result
con.execute("SELECT * FROM points_clustered LIMIT 5").df()
x1 | x2 | y | labels
-- | -- | -- | --
0.203493 | 0.714086 | 0.788261 | -1
0.279848 | 0.223594 | 0.163997 | 0
0.734297 | 0.219713 | 0.269225 | 0
0.809632 | 0.852975 | 0.849835 | -1
0.745970 | 0.304509 | 0.206545 | 0
```

If you're already using e.g., scikit-learn or hdbscan in Python, you could use cuML's [zero code change UX](https://docs.rapids.ai/api/cuml/stable/cuml-accel/).

**Comment by fhk** [comment]
bump :)
